persipubsub/subscriber.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+"""Receive messages persistent from publisher to subscriber."""
+
+import contextlib
+import datetime
+import pathlib
+import time
+from typing import Generator, Optional, Union
+
+import icontract
+import lmdb  # pylint: disable=unused-import
+
+import persipubsub.queue
+
+# pylint: disable=protected-access
+
+
+class Subscriber:
+    """
+    Create Subscriber ready to receive messages.
+
+    :ivar identifier: subscriber ID
+    :vartype pub_id: str
+    :ivar queue: from which messages are received
+    :vartype queue: persipubsub.queue.Queue
+    """
+
+    def __init__(self) -> None:
+        """Initialize class object."""
+        self.identifier = None  # type: Optional[str]
+        self.queue = None  # type: Optional[persipubsub.queue._Queue]
+
+    def init(self, identifier: str, path: Union[pathlib.Path, str]) -> None:
+        """
+        Initialize.
+
+        :param identifier: unique subscriber id
+        :param path: path to the queue
+        """
+        self.identifier = identifier
+        assert isinstance(self.identifier, str)
+        self.queue = persipubsub.queue._Queue()  # pylint: disable=protected-access
+        self.queue.init(path=path)
+        assert isinstance(self.queue, persipubsub.queue._Queue)
+
+    def __enter__(self) -> 'Subscriber':
+        """Enter the context and give the sub prepared in the constructor."""
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        """Leave the context."""
+
+    @icontract.require(lambda timeout: timeout > 0)
+    @icontract.require(lambda retries: retries > 0)
+    @contextlib.contextmanager
+    def receive(self, timeout: int = 60, retries: int = 10) -> Generator:
+        """
+        Receive messages from the publisher.
+
+        :param timeout: time waiting for a message. If none arrived until the
+            timeout then None will be returned. (secs)
+        :param retries: number of tries to check if a msg arrived in the queue
+        """
+        msg = None
+        end = int(datetime.datetime.utcnow().timestamp()) + timeout
+        assert isinstance(self.queue, persipubsub.queue._Queue)
```

Comment:
Adding all this asserts was troublesome. I looked into you aproach, but I'm not sure if that would solve the problem. The problem is that assert isinstance is needed for mypy because the variable is initialized as type: Optional[...] in __init__() and then initialized in init() as type: ... and I didn't seem a way to update the type annotation from Optional[...] to ...
As alternative approach I found: https://github.com/zulip/zulip/blob/master/mypy.ini
In their project they just deactivate the mypy --strict flag for some methods. Although I guess this will make the flag useless in the first place. So I don't think that's the solution.

Reply:
Interesting chapters in mypy documentation https://mypy.readthedocs.io/en/latest/kinds_of_types.html#optional-types-and-the-none-type:

- Union types
- Optional and the None type
- Disabling optional checking